### PR TITLE
Static build reimplemented in more portable way

### LINKED ISSUE #54: Static build fails for dwarfsextract
Static build fails for dwarfsextract due to unresloved references from libarchive.a
I guess it happens because libarchive from focal binary package is statically linked to more dependencies that are specified in _static_link.sh_  (as per _https://launchpad.net/ubuntu/focal/+source/libarchive_)

So _dwarfs_ static build requires either custom _libarchive.a_ that matches supported formats or larger set of dependencies 

`[430/431] Linking CXX executable dwarfsextract
FAILED: dwarfsextract 
: && /bin/bash /mnt/d/Projects/5.Projects/tebako/deps/src/_dwarfs/cmake/static_link.sh dwarfsextract CMakeFiles/dwarfsextract.dir/src/dwarfsextract.cpp.o && :
...
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_write_set_format_xar.o): in function `compression_code_bzip2':
(.text+0xc5d): undefined reference to `BZ2_bzCompress'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_write_set_format_xar.o): in function `xar_options':
(.text+0x11b1): undefined reference to `lzma_cputhreads'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_write_set_format_xar.o): in function `compression_end_bzip2':
(.text+0x17ec): undefined reference to `BZ2_bzCompressEnd'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_write_set_format_xar.o): in function `xar_compression_init_encoder':
(.text+0x20fa): undefined reference to `BZ2_bzCompressInit'
/usr/bin/ld: (.text+0x2261): undefined reference to `lzma_stream_encoder_mt'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_cryptor.o): in function `aes_ctr_encrypt_counter':
(.text+0x52): undefined reference to `nettle_aes_set_encrypt_key'
/usr/bin/ld: (.text+0x6d): undefined reference to `nettle_aes_encrypt'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_cryptor.o): in function `pbkdf2_sha1':
(.text+0x2a1): undefined reference to `nettle_pbkdf2_hmac_sha1'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha512final':
(.text+0x11): undefined reference to `nettle_sha512_digest'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha384update':
(.text+0x32): undefined reference to `nettle_sha512_update'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha512init':
(.text+0x49): undefined reference to `nettle_sha512_init'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha384final':
(.text+0x71): undefined reference to `nettle_sha384_digest'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha384init':
(.text+0x89): undefined reference to `nettle_sha384_init'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha256final':
(.text+0xb1): undefined reference to `nettle_sha256_digest'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha256update':
(.text+0xd2): undefined reference to `nettle_sha256_update'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha256init':
(.text+0xe9): undefined reference to `nettle_sha256_init'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha1final':
(.text+0x111): undefined reference to `nettle_sha1_digest'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha1update':
(.text+0x132): undefined reference to `nettle_sha1_update'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha1init':
(.text+0x149): undefined reference to `nettle_sha1_init'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_ripemd160final':
(.text+0x171): undefined reference to `nettle_ripemd160_digest'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_ripemd160update':
(.text+0x192): undefined reference to `nettle_ripemd160_update'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_ripemd160init':
(.text+0x1a9): undefined reference to `nettle_ripemd160_init'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_md5final':
(.text+0x1d1): undefined reference to `nettle_md5_digest'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_md5update':
(.text+0x1f2): undefined reference to `nettle_md5_update'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_md5init':
(.text+0x209): undefined reference to `nettle_md5_init'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_digest.o): in function `__archive_nettle_sha512update':
(.text+0x232): undefined reference to `nettle_sha512_update'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_hmac.o): in function `__hmac_sha1_init':
(.text+0x92): undefined reference to `nettle_hmac_sha1_set_key'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_hmac.o): in function `__hmac_sha1_final':
(.text+0x4d): undefined reference to `nettle_hmac_sha1_digest'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_hmac.o): in function `__hmac_sha1_update':
(.text+0x6e): undefined reference to `nettle_hmac_sha1_update'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_write_set_format_7zip.o): in function `compression_code_bzip2':
(.text+0x32d): undefined reference to `BZ2_bzCompress'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_write_set_format_7zip.o): in function `compression_end_bzip2':
(.text+0xc0c): undefined reference to `BZ2_bzCompressEnd'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libarchive.a(archive_write_set_format_7zip.o): in function `_7z_compression_init_encoder':`

### PR DESCRIPTION
**Pull request by maxirmx**
Hello 
This PR includes dwarfs static build reimplemented in more portable way that shall fix #54 and similar issues that may exist in the filed.
CMakeLists.txt includes detailed explanantion of the changes that I suggest.

Thank you

cc: @ronaldtse

This is a contribution from [Metanorma](https://www.metanorma.org) ([GitHub](https://github.com/metanorma)), a product of [Ribose](https://www.ribose.com) ([GitHub](https://github.com/riboseinc)).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,9 @@
 /.gdb_history
 /include/dwarfs/version.h
 /src/dwarfs/version.cpp
+
+# VS build environment
+.vs/
+out/
+CMakeSettings.json
+fbthrift/
```
Comment (mhx) [inline .gitignore:16]: Why was this added?

**Comment by mhx** [comment]
Thanks for the PR!

I need to take a closer look and play with it when I've got access to my VMs again next week.

File: CMakeLists.txt [hunk 2]
```diff
@@ -620,12 +632,96 @@ foreach(tgt ${BINARY_TARGETS})
...
+    target_link_libraries(${tgt}
+      ${_LIBFMT}          ${_LIBDC}           ${_LIBGLOG}         ${_LIBGFLAGS}                
+      _LIBARCHIVE         ${_LIBEVENT}        ${_LIBACL}          
+      ${_LIBSSL}          ${_LIBCRYPTO}       ${_LIBPTHREAD}      ${_LIBDL}
+      ${_LIBZ}            ${_LIBC}            ${_LIBM}            ${_LIBRT}
+      gcc_eh              ${_LIBUNWIND}             
+    )
+  endforeach()
+endif(STATIC_BUILD_DO_NOT_USE)
+# icui18n icuuc icudata
```
Comment (mhx) [inline CMakeLists.txt:724]: What's this comment about?

Comment (maxirmx) [reply CMakeLists.txt:724]: It used to be a list of libraries that were required by libarchieve under certain circumstances.
Just forgot to delete it.

File: .gitignore (hunk 1 above)
Comment (maxirmx) [reply .gitignore:16]: .vs/**,   out/**, CMakeSettings.json  are configuration files and build artefacts generated by Visual Studio. This was added to support this build environment but it is not that important, of course.

fbthrift was added because there are some file generated in-source during build.

File: CMakeLists.txt (hunk 2 above)
Comment (maxirmx) [reply CMakeLists.txt]: Removed

**Comment by mhx** [comment]
This is pretty cool! It's *almost* there; tests aren't linking properly, but I've got a fix for that already. Thanks!
